src/main/kotlin/me/partlysanestudios/partlysaneskies/features/dungeons/TerminalWaypoints.kt

```diff
@@ -0,0 +1,181 @@
+        val playerBoundingBox = getCurrentBoundingBox(playerPoint3d)
+
+
+
+        for (pointPair in getAllInactivePos()) {
+            val point = pointPair.key
+            val type = pointPair.value
+            if (playerBoundingBox?.isInRange(point) != true) {
+                continue
+            }
+
+            val puzzleFound = findPuzzle(point)
+            if (puzzleFound == null) {
+                val puzzle = F7Puzzle(Point3d(point.toBlockPosInt()), type)
+                terminals.add(puzzle)
+            } else {
+                if (!puzzleFound.active) {
+                    puzzleFound.active = false
+                }
+            }
+        }
+
+        for (pointPair in getAllActivePos()) {
+            val point = pointPair.key
+            val type = pointPair.value
+            if (playerBoundingBox?.isInRange(point) != true) {
+                continue
+            }
+
+            val puzzleFound = findPuzzle(point)
+            if (puzzleFound == null) {
+                val puzzle = F7Puzzle(Point3d(point.toBlockPosInt()), type)
+
+                puzzle.active = true
+                terminals.add(puzzle)
+            } else {
+                if (!puzzleFound.active) {
+                    puzzleFound.active = true
+                }
+            }
+        }
+
+        for (puzzle in terminals) {
+            if (playerBoundingBox?.isInRange(puzzle.pos) != true) {
+                continue
+            }
+            val waypoint = if (puzzle.active) {
+                Waypoint(
+                    puzzle.type.displayName,
+                    puzzle.pos.toBlockPosInt(),
+                    fillColor = Color.green.applyOpacity(50),
+                    outlineColor = Color.green.applyOpacity(125)
+                )
+            } else {
+                Waypoint(
+                    puzzle.type.displayName,
+                    puzzle.pos.toBlockPosInt(),
+                    fillColor = Color.red.applyOpacity(50),
+                    outlineColor = Color.red.applyOpacity(125)
+                )
+            }
+
+            event.pipeline.add(waypoint)
+        }
+    }
+
+    fun getAllActivePos(): Map<Point3d, F7Puzzle.Type> {
+        val armorStands = getAllArmorStands()
+        val activeDevicePoses = HashMap<Point3d, F7Puzzle.Type>()
+
+        for (armorStand in armorStands) {
+            val name = armorStand.displayName.unformattedText
+            for (type in F7Puzzle.Type.entries) {
+                if (!name.contains(type.activeArmorStandName)) {
+                    continue
+                }
+
+                val pos = Point3d(armorStand.posX, armorStand.posY + 1, armorStand.posZ)
+                activeDevicePoses[pos] = type
+            }
+        }
+
+        return activeDevicePoses
+    }
+
+
+    @SubscribePSSEvent
+    fun onDungeonStart(event: DungeonStartEvent) { // Yes I wrote the entire event system just so that I wouldn't have to call a chat event here
```

Comment:
Feel free to remove the comment since we now use it multiple times

Reply:
The reason i wrote the event system was purely motivated by that like right there

Reply:
lmao